tools/src/main/java/org/apache/kafka/tools/StorageTool.java

```diff
@@ -0,0 +1,555 @@
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metadata.FeatureLevelRecord;
+import org.apache.kafka.common.metadata.UserScramCredentialRecord;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.security.scram.internals.ScramFormatter;
+import org.apache.kafka.common.security.scram.internals.ScramMechanism;
+import org.apache.kafka.common.utils.Exit;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.metadata.bootstrap.BootstrapDirectory;
+import org.apache.kafka.metadata.bootstrap.BootstrapMetadata;
+import org.apache.kafka.metadata.properties.MetaProperties;
+import org.apache.kafka.metadata.properties.MetaPropertiesEnsemble;
+import org.apache.kafka.metadata.properties.MetaPropertiesVersion;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.server.common.MetadataVersion;
+import org.apache.kafka.server.common.ProcessRole;
+import org.apache.kafka.server.config.ReplicationConfigs;
+import org.apache.kafka.server.config.ServerLogConfigs;
+import org.apache.kafka.storage.internals.log.LogConfig;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+import static net.sourceforge.argparse4j.impl.Arguments.append;
+import static net.sourceforge.argparse4j.impl.Arguments.store;
+import static net.sourceforge.argparse4j.impl.Arguments.storeTrue;
+
+public class StorageTool {
+    // TODO delete these constants when they are migrated from KafkaConfig
+    public static final String METADATA_LOG_DIR_CONFIG = "metadata.log.dir";
+    public static final String UNSTABLE_METADATA_VERSIONS_ENABLE_CONFIG = "unstable.metadata.versions.enable";
+    public static final String PROCESS_ROLES_CONFIG = "process.roles";
+    
+    public static void main(String... args) {
+        try {
+            
+            Namespace namespace = parseArguments(args);
+            String command = namespace.getString("command");
+            Optional<LogConfig> logConfig = Optional.ofNullable(namespace.getString("config")).map(p -> {
+                try {
+                    return new LogConfig(Utils.loadProps(p));
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+            executeCommand(namespace, command, logConfig);
```

Comment:
It's not really nice that we cast the configs to `LogConfig` but then have to rely on `originals()` to access non LogConfig configs.

Reply:
Ok. At least this allows to test the migration without generating too many conflicts with the work about migrating KafkaConfig.